### PR TITLE
Fixed the real cause of NPEs related to FormController.

### LINKED ISSUE #2850: NullPointerException in FormHierarchyActivity#218
#### Software and hardware versions 
Collect v1.19.0 Android 4 and 5

#### Problem description
https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/5c57cfdff8b88c2963958034?time=last-seven-days&sessionId=5C57CFDA002E000120AB5CFCA96EFFD4_DNE_0_v2

`Fatal Exception: java.lang.NullPointerException
Attempt to invoke virtual method 'boolean org.javarosa.core.model.FormIndex.isBeginningOfFormIndex()' on a null object reference
keyboard_arrow_up
arrow_right
org.odk.collect.android.activities.FormHierarchyActivity.updateOptionsMenu (FormHierarchyActivity.java:218)
org.odk.collect.android.activities.FormHierarchyActivity.onPrepareOptionsMenu (FormHierarchyActivity.java:206)
android.app.Activity.onPreparePanel (Activity.java:2939)
android.support.v4.app.FragmentActivity.onPrepareOptionsPanel (FragmentActivity.java:559)
android.support.v4.app.FragmentActivity.onPreparePanel (FragmentActivity.java:547)`

#### Other information 
It's not a new issue, not very common (4users affected, 5 events reported)

@cooperka could you take a looks since it's related to your latest changes?

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #2850 

#### What has been done to verify that this works as intended?
It's difficult to test. I wasn't able to reproduce the issue so I just hardcoded null value to call this block:
https://github.com/opendatakit/collect/compare/master...grzesiek2010:fc_crashes?expand=1#diff-03ff45091550285dd88b56927010c457R453

#### Why is this the best possible solution? Were any other approaches considered?
It's a long-standing issue that has been causing many NPEs like:
https://github.com/grzesiek2010/collect/commit/c1f80c1c5ca8ecd53b7145b0a1e638c3ec578927
https://github.com/grzesiek2010/collect/commit/d4abc86b37ec6f2fc6c967fb7221595068d15938
https://github.com/grzesiek2010/collect/commit/560c9c50cd0e4850b8ff2d83a93da8332256490c
https://github.com/grzesiek2010/collect/commit/2296c5e3375ea2f881ba7cc9d081accd66662ef7
https://github.com/grzesiek2010/collect/commit/eb0d141925cf97bdf7a676e159ccef34c8e56e56
https://github.com/grzesiek2010/collect/commit/d14ed919130ce8f9e0c6ef9e1438b812f6996a71
and many others.

We have been adding just workarounds so far but now we have a real fix. The problem was that we don't show any dialog (we don't block the UI) when a form is reloaded so a user can swipe or open form hierarchy what causes NPEs.

Can we get rid of all those null checks we added in commits mentioned above? I'm pretty sure we can but to be 100% I added the second commit which will log a special event if a null form controller takes place in any other case that the one I handle here.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Since I wasn't able to reproduce the issue it's going to be difficult for our QAs as well... 
I added a special commit just for testers which allows setting formController value. Just press a volume button (up or down) no matter if it's `FormEntryActivity` or `FormHierarchyActivity`.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 1]
```diff
@@ -446,17 +446,13 @@ private void loadForm() {
...
+                if (getFormController(true) != null) {
                     refreshCurrentView();
                 } else {
                     Timber.w("Reloading form and restoring state.");
-                    // we need to launch the form loader to load the form
-                    // controller...
-                    formLoaderTask = new FormLoaderTask(instancePath,
-                            startingXPath, waitingXPath);
-                    // TODO: this doesn' work (dialog does not get removed):
-                    // showDialog(PROGRESS_DIALOG);
```
Comment (cooperka) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:458]: I encounter this "TODO" issue when hard-coding this execution path (dialog does not get removed, it just remains spinning until I press the cancel button).

**Review by cooperka — COMMENTED** [review]
Nice job tracking this down @grzesiek2010. This looks like it might be the right fix, but like you mentioned, there's no good way to verify since it's still unclear why `formController` is `null` in the first place.

If you're able to fix the infinite spinner bug that would be great, though it would obviously be rare that a user would encounter it in the first place.

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: That's not true (the comment) I tested it and the dialog is removed properly. It's a very old code so probably it was something wrong back then.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:458]: I fixed that problem using DialogFragment instead.

**Review by cooperka — APPROVED** [review]
Awesome. Volume button helps a lot for testing.

I noticed it resets the form index when reloading formController (e.g. I'm on question 5, make formController null, leave and come back, now it reloads and I'm on question 1) which could be an inconvenience on devices with this crash condition that are eternally low on memory. I can imagine a situation where there's an external app to generate question values and every time you open that app Android kills Collect and you have to start from question 1.

Again, I couldn't reproduce this crash because for me formController remains non-null in every scenario I could think of to test (except when forcing it with the volume button). It's rare, and reloading is certainly better than crashing. I personally think this PR is fine to merge as-is unless a fix is easy and safe and testable.

**Comment by grzesiek2010** [comment]
> I can imagine a situation where there's an external app to generate question values and every time you open that app Android kills Collect and you have to start from question 1.

yes, it would be a big problem but it's not something that happens constantly and no one complained about it on the forum so I think it's ok.

Thanks for your review.

**Comment by cooperka** [comment]
> it's not something that happens constantly and no one complained about it on the forum

Just to be clear, this would happen constantly in the hypothetical described situation, and it's a new issue after this PR, not a pre-existing issue. But assuming the crash itself that this PR fixes is rare, then this issue is **much** better than a crash.

**Comment by mmarciniak90** [comment]
@grzesiek2010 please look at my test results

* 'Do not keep activities' turned ON on all device
* verified on 'All widgets' form
* verified 'Nigeria Wards' form to case 3

|ANDROID |CASE 1 | CASE 2| CASE 3|
| ---| ---|---| ---|
| | 1. User sees form hierarchy<br>2. User clicks on **device back button** |   1. User sees form hierarchy<br>2. User **changes device orientation**| 1. User is on some question (not first)<br>2. User changes device orientation|
| | **RESULT ON MASTER**| **RESULT ON MASTER**|**RESULT ON MASTER**|
|master on Android 6.0 |The view is reloaded and form hierarchy is still visible. <br><br>**I checked also this case with the second form. I was able to click on the Hierarchy icon which causes crash** | Crash| User is moved to first question. Added answers are cleared. Loading dialog **is not** visible. <br><br>**I was able to click on the Hierarchy icon which causes crash**|
| |**RESULT ON APK WITH FIX** | **RESULT ON APK WITH FIX**|**RESULT ON APK WITH FIX**|
|Android 8.1| Form loading dialog is visible and question appears | Form loading dialog is visible and question appears| The current page changes orientation. <br>* **App crashes if I try go to next/prev question.** <br> * **Main view appears If I clicked on the Hierarchy icon.** <br>* **Expected dialog appeared when I opened General Setting and returned to form** <br> * **App crashes if I click on the save icon.** |
|Android 7.0|Form loading dialog is visible and question appears |Form loading dialog is visible and question appears | User is moved to first question. Added answers are cleared. Loading dialog **is** visible. There is no way to click on the Hierarchy icon.|
|Android 6.0| Form loading dialog is visible and question appears | Crash |User is moved to first question. Added answers are cleared. Loading dialog **is** visible. There is no way to click on the Hierarchy icon.|
|Android 5.1| Form loading dialog is visible and question appears | Crash |User is moved to first question. Added answers are cleared. Loading dialog **is** visible. There is no way to click on the Hierarchy icon.|
|Android 4.4|Form loading dialog is visible and question appears | Crash |User is moved to first question. Added answers are cleared. Loading dialog **is** visible. There is no way to click on the Hierarchy icon.|

**Comment by mmarciniak90** [comment]
I retested issue after small changes.
Here are the current test results.

|ANDROID |CASE 1 | CASE 2| CASE 3|
| ---| ---|---| ---|
| | 1. _User sees form hierarchy_<br>2. _User clicks on device back button_ |   1. _User sees form hierarchy_<br>2. _User changes device orientation_| 1. _User is on some question (not first)<br>2. User changes device orientation_|
| |**RESULT**|**RESULT**|**RESULT**|
|Android 8.1|Form loading dialog is visible and question appears|Form loading dialog is visible and question appears|The current page changes orientation. No crash|
|Android 7.0|Form loading dialog is visible and question appears|Form loading dialog is visible and question appears|The hierarchy view appears|
|Android 6.0|Form loading dialog is visible and question appears|Form loading dialog is visible and question appears|The hierarchy view appears|
|Android 5.1|Form loading dialog is visible and question appears|Form loading dialog is visible and question appears|The hierarchy view appears|
|Android 4.4|Form loading dialog is visible and question appears|Form loading dialog is visible and question appears|The hierarchy view appears|

@grzesiek2010 In my opinion, results are good on all Android versions. Behavior on Android 8.1 for case 3 is different but I did not notice crash so I think that it is still ok. 

Probably, there is no option to retest this bug during regression testing, so I want to be sure that tests were executed correctly.
@grzesiek2010 do you have any suggestions?

**Comment by grzesiek2010** [comment]
Yes it was difficult to track down and it's difficult to test but you did well. Those result's look fine. You can add `behavior verified` label, then I'll get rid of the commit for testing and we can marge.

**Comment by mmarciniak90** [comment]
Thanks, @grzesiek2010 :+1: 

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Comment by grzesiek2010** [comment]
Ok, I removed the last commit (for testing) so we can merge.

**Review by shobhitagarwal1612 — APPROVED** [review]

**Pull request merged**